ACPs/176-dynamic-evm-gas-limit-and-price-discovery-updates/README.md

```diff
@@ -0,0 +1,158 @@
+
+$$q_{desired} = D \cdot ln(\frac{T}{P})$$
+
+Note that since $q_{desired}$ is only used locally and can be different for each node, it is safe for implementations to approximate the value of $ln(\frac{T}{P})$, and round the resulting value to the nearest integer.
+
+When building a block, builders can calculate their next preferred value for $q$ according to:
+
+  ```python
+  # Calculates a node's new desired value for q given for a given block
+  def calc_next_q(q_current: int, q_desired: int, max_change: int) -> int:
+    if q_desired > q_current:
+        return q_current + min(q_desired - q_current, max_change)
+    else:
+        return q_current - min(q_current - q_desired, max_change)
+  ```
+
+As $q$ is updated after the execution of transactions within the block, $T$ is also updated such that $T = P \cdot e^{q/D}$ at all times. As the value of $T$ adjusts, the value of $R$ (capacity added per second) is also updated such that:
+
+$$R = T \cdot 2$$
+
+This ensures that the gas price remains equally reactive in either direction depending on how far the actual gas consumption rate is from the target, whether it is above or below.
+
+The value of $C$ must also adjust proportionately, so we set:
+
+$$C = R \cdot 10$$
+
+This means that the maximum stored gas capacity would be reached after 10 seconds where no blocks have been accepted.
+
+In order to keep roughly constant the time it takes for the gas price to double at sustained maximum network capacity usage, the value of $K$ used in the gas price determination mechanism must be updated proportionatly to $T$. When updating $T$ from $T_{n}$ to $T_{n+1}$, we update $K$ such that:
+
+$$K_{n+1} = K_{n} \cdot \frac{T_{n+1}}{T_{n}}$$
+
+In order to have the gas price not be directly impacted by the change in $K$, we also update $x$ (excess gas consumption) proportionally. When updating $x$ after executing a block, instead of setting $x = x + G$ as specified in ACP-103, we set:
+
+$$x_{n+1} = (x + G) \cdot \frac{K_{n+1}}{K_{n}}$$
+
+Note that the value of $q$ (and thus also $T$, $R$, $C$, $K$, and $x$) are updated **after** the execution of block $b$, which means they only take effect in determining the gas price of block $b+1$. The change to each of these values in block $b$ does not effect the gas price for transactions included in block $b$ itself.
+
+Allowing block builders to adjust the target gas consumption rate in blocks that they produce makes it such that the effect target gas consumption rate should converge over time around the stake-weighted average value set by validators of the network. This is because the number of blocks each validator produces is proportional to their stake weight. This means that an individual validator's effect on the resulting target gas consumption for the network is proportional to their stake weight.
```

Comment:
> the effect target gas consumption rate should converge over time around the stake-weighted average value set by validators of the network.

I don't think this is quite accurate. I think the effective target gas consumption rate should converge to the point where 50% of voting stake wants it decreased and 50% of voting stake wants it increased.

If 49% of the network wanted the target gas consumption to go to 0 and 51% of the network wanted the target gas consumption to go to $x$, then the target would move towards $x$ (albeit slowly) over time.

Reply:
Interesting, I had not realized that...so a stake weight equilibrium rather than average in a way.

Reply:
I updated it now, and removed the final sentence for conciseness.